### PR TITLE
Improve exception handling and rename IonElectrolyteException to IonElementException

### PR DESCRIPTION
**Pull request by dlurton**
Implements #14

- Rename IonElectrolyteException to IonElementException
- Add IonElementLoaderException and throw it from IonElementLoaderImpl
- Add IonElementConstraintException and throw it from AnyIonElement

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/com/amazon/ionelement/api/AnyElement.kt [hunk 1]
```diff
@@ -49,8 +49,8 @@ import java.math.BigInteger
  * - An expectation of the Ion type of the given element
  * - An expectation of the nullability of the given element
  *
- * If either of these expectations is violated an [IonElectrolyteException] is thrown.  If the given element
- * has a [IonLocation] in its metadata, it included with the [IonElectrolyteException] which can be used to
+ * If either of these expectations is violated an [IonElementException] is thrown.  If the given element
+ * has a [IonLocation] in its metadata, it included with the [IonElementException] which can be used to
```
Comment (therapon) [inline src/com/amazon/ionelement/api/AnyElement.kt:53]: it is included

File: src/com/amazon/ionelement/api/IonElementException.kt [hunk 2]
```diff
@@ -15,16 +15,25 @@
...
 package com.amazon.ionelement.api
 
-import com.amazon.ionelement.api.IonLocation
-import com.amazon.ionelement.api.locationToString
-
 /**
- * Exception for IonElectrolyte.
- *
- * Includes the [IonLocation] in the message if the [location] property if it was included at construction time.
+ * Base exception which includes the [location] in the message if the if it was included at construction time.
```
Comment (therapon) [inline src/com/amazon/ionelement/api/IonElementException.kt:19]: remove `the if`

File: src/com/amazon/ionelement/api/IonElementLoader.kt [hunk 3]
```diff
@@ -17,11 +17,18 @@ package com.amazon.ionelement.api
 
 import com.amazon.ion.IonReader
 
+/**
+ * Provides several functions for loading [IonElement] instances.
+ *
+ * All functions wrap any [com.amazon.ion.IonException] in an instance of [IonElementLoaderException], including the
+ * current [IonLocation] if if one is available.  Note that depending on the state of the [IonReader], a location
```
Comment (therapon) [inline src/com/amazon/ionelement/api/IonElementLoader.kt:24]: if if => if

File: src/com/amazon/ionelement/api/IonUtils.kt [hunk 4]
```diff
@@ -52,12 +52,12 @@ fun IonValue.toIonElement(): AnyElement =
...
+/** Throws an [IonElementException], including the [IonLocation] (if available). */
+internal fun constraintError(blame: IonElement, description: String): Nothing {
+    constraintError(blame.metas, description)
 }
 
-/** Throws an [IonElectrolyteException], including the [IonLocation] (if available). */
-fun ionError(metas: MetaContainer, description: String): Nothing {
-    throw IonElectrolyteException(metas.location, description)
+/** Throws an [IonElementException], including the [IonLocation] (if available). */
+internal fun constraintError(metas: MetaContainer, description: String): Nothing {
```
Comment (therapon) [inline src/com/amazon/ionelement/api/IonUtils.kt:61]: This used to be an exposed function or is my Kotlin knowledge failing me again?

Comment (dlurton) [reply src/com/amazon/ionelement/api/IonUtils.kt:61]: it was part of the API previously, but I doubt anyone is using it.  Even so, I am breaking so many other things with `v0.2.0` I thought:  "what's one more"?  

In any case, I don't know if it's all that useful to provide an API to throw an `IonElementConstraintException` from client code and I erred on the side of a smaller API.  I could remove the `internal`, if you think it's helpful.

Comment (therapon) [reply src/com/amazon/ionelement/api/IonUtils.kt:61]: No it's fine as is

File: src/com/amazon/ionelement/api/AnyElement.kt (hunk 1 above)
Comment (dlurton) [reply src/com/amazon/ionelement/api/AnyElement.kt:53]: done

File: src/com/amazon/ionelement/api/IonElementException.kt (hunk 2 above)
Comment (dlurton) [reply src/com/amazon/ionelement/api/IonElementException.kt:19]: done

File: src/com/amazon/ionelement/api/IonElementLoader.kt (hunk 3 above)
Comment (dlurton) [reply src/com/amazon/ionelement/api/IonElementLoader.kt:24]: done

**Pull request merged**